src/wp-includes/class-wp-theme.php

```diff
@@ -1993,16 +2001,38 @@ private function set_pattern_cache( array $patterns ) {
 			'version'  => $this->get( 'Version' ),
 			'patterns' => $patterns,
 		);
-		wp_cache_set( 'wp_theme_patterns_' . $this->stylesheet, $pattern_data, 'theme_files' );
+
+		/**
+		 * Filters whether to cache theme files persistently.
+		 *
+		 * @since 6.6.0
+		 *
+		 * @param bool   $cache_expiration Expiration time for the cache. Setting a value to `false` would bypass caching. Default `WP_Theme::$cache_expiration`.
+		 * @param string $context          Additional context for better cache control.
+		 */
+		$cache_expiration = apply_filters( 'wp_cache_theme_files_persistently', self::$cache_expiration, 'theme_block_patterns' );
+
+		if ( false !== $cache_expiration ) {
+			if ( wp_using_ext_object_cache() ) {
+				wp_cache_set( 'wp_theme_patterns_' . $this->stylesheet, $pattern_data, 'theme_files', $cache_expiration );
+			} else {
+				set_site_transient( 'theme_files_wp_theme_patterns_' . $this->stylesheet, $pattern_data, $cache_expiration );
+			}
+		}
```

Comment:
Per https://github.com/WordPress/wordpress-develop/pull/6137/files#r1520119762, I think the filter should be limited to only control cache expiration which is what we had originally discussed in the ticket:
```suggestion
		if ( wp_using_ext_object_cache() ) {
			wp_cache_set( 'wp_theme_patterns_' . $this->stylesheet, $pattern_data, 'theme_files', $cache_expiration );
		} else {
			set_site_transient( 'theme_files_wp_theme_patterns_' . $this->stylesheet, $pattern_data, $cache_expiration );
		}
```

Reply:
Done.